src/render_systems.rs

```diff
@@ -334,3 +339,229 @@ pub fn queue_pipelines_system(
+                            callback: callback.clone(),
+                            rect: paint_callback.rect,
+                        },
+                    ));
+
+                    data.draw_commands.push(DrawCommand {
+                        primitive: DrawPrimitive::PaintCallback(PaintCallbackDraw {
+                            callback,
+                            rect: paint_callback.rect,
+                        }),
+                        clip_rect,
+                    });
+                    continue;
+                }
+            };
+
+            data.vertex_data
+                .extend_from_slice(cast_slice::<_, u8>(mesh.vertices.as_slice()));
+            data.index_data
+                .extend(mesh.indices.iter().map(|i| i + index_offset));
+            index_offset += mesh.vertices.len() as u32;
+
+            let texture_handle = match mesh.texture_id {
+                egui::TextureId::Managed(id) => EguiTextureId::Managed(*main_entity, id),
+                egui::TextureId::User(id) => EguiTextureId::User(id),
+            };
+
+            data.draw_commands.push(DrawCommand {
+                primitive: DrawPrimitive::Egui(EguiDraw {
+                    vertices_count: mesh.indices.len(),
+                    egui_texture: texture_handle,
+                }),
+                clip_rect,
+            });
+        }
+
+        if data.vertex_data.len() > data.vertex_buffer_capacity {
+            data.vertex_buffer_capacity = round_up_to_pow2(data.vertex_data.len());
+            data.vertex_buffer = Some(render_device.create_buffer(&BufferDescriptor {
+                label: Some("egui vertex buffer"),
+                size: data.vertex_buffer_capacity as BufferAddress,
+                usage: BufferUsages::COPY_DST | BufferUsages::VERTEX,
+                mapped_at_creation: false,
+            }));
+        }
+
+        let index_data_size = data.index_data.len() * std::mem::size_of::<u32>();
+        if index_data_size > data.index_buffer_capacity {
+            data.index_buffer_capacity = round_up_to_pow2(index_data_size);
+            data.index_buffer = Some(render_device.create_buffer(&BufferDescriptor {
+                label: Some("egui index buffer"),
+                size: data.index_buffer_capacity as BufferAddress,
+                usage: BufferUsages::COPY_DST | BufferUsages::INDEX,
+                mapped_at_creation: false,
+            }));
+        }
+
+        let (vertex_buffer, index_buffer) = match (&data.vertex_buffer, &data.index_buffer) {
+            (Some(vertex), Some(index)) => (vertex, index),
+            _ => {
+                continue;
+            }
+        };
+
+        render_queue.write_buffer(vertex_buffer, 0, &data.vertex_data);
+        render_queue.write_buffer(index_buffer, 0, cast_slice(&data.index_data));
+    }
+}
+
+fn round_up_to_pow2(size: usize) -> usize {
+    if size.is_power_of_two() {
+        size
+    } else {
+        size.next_power_of_two()
+    }
+}
```

Comment:
This function seems to be redundant: https://doc.rust-lang.org/std/primitive.usize.html#method.next_power_of_two
```
assert_eq!(2usize.next_power_of_two(), Some(2));
```